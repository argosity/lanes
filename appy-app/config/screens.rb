
Lanes::Screen.define "base" do | screen |
    screen.description = "Base Screen for Appy App"
    screen.icon        = ""
    screen.group_id    = ""
    screen.model_class = ""
    screen.view_class  = "AppyApp.Screens.Base"
end
