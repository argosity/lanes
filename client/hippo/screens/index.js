import { observable, computed } from 'mobx';
import { map, filter, intersection } from 'lodash';
import Group from './group';
import Config from '../config';

const Screens = observable({

    all: observable.map(),

    get activeGroups() {
        const gids = Group.enabled_group_ids || map(Group.all, 'id');
        const groups = map(gids, gid => Group.forId(gid));
        return filter(groups, group => intersection(group.screens, this.active).length);
    },

    @computed get active() {
        return map(Config.screen_ids, id => this.all.get(id));
    },

    reset() {
        Config.screen_ids.clear();
    },

});

export default Screens;
