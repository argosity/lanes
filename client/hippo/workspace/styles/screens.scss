.screen {
    padding: 5px;
    width: 100%;

    > .first-run {
        font-size: 1.3rem;
        li + li {
            margin-top: 1rem;
        }
        .help {
            margin-top: 2rem;
            p { text-align: right; }
        }
    }
}
