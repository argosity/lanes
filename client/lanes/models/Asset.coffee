IMAGES = [
    "image/png", "image/jpg", "image/gif"
]

IS_IMAGE = (content_type) ->
    !!(content_type && IMAGES.indexOf(content_type) isnt -1)

class Lanes.Models.Asset extends Lanes.Models.Base

    props:
        id:        'integer'
        order:     'integer'
        file_data: 'object'

    session:
        data:     'string'
        blob:     'object'
        owner:    'object'
        parent:   'object'
        metadata: 'object'
        parent_association: 'string'

    derived:
        original_url:
            deps: ['data', 'file_data'], fn: ->
                if @data then @data else @urlFor('original')
        medium_url:
            deps: ['data', 'file_data'], fn: ->
                if @data then @data else @urlFor('medium')
        thumbnail_url:
            deps: ['data', 'file_data'], fn: ->
                if @data then @data else @urlFor('thumbnail')

        isPresent:
            deps: ['data', 'file_data'], fn: ->
                @data || !_.isBlank(@file_data)

        hasImage:
            deps: ['data', 'file_data'], fn: ->
                (@data && IS_IMAGE(@metadata.content_type)) ||
                    (@file_data && IS_IMAGE(@file_data.original?.metadata.mime_type))
    events:
        'change:blob': 'onBlobChange'

    initialize: ->
        @on('change:parent', =>
            oldParent = _this. previousAttributes().parent
            @stopListening(oldParent, 'save', @save) if oldParent
            @listenTo(@parent, 'save', @save)
        )

    baseUrl: ->
        Lanes.config.api_path + '/asset'

    urlFor: (type) ->
        data = @file_data[type]
        if data then "#{@baseUrl()}/#{data.id}" else undefined

    onBlobChange: ->
        if @blob
            @metadata = { content_type: @blob.type }
            reader = new FileReader()
            reader.onloadend = (ev) =>
                @data = reader.result if ev.type is 'loadend'
            reader.readAsDataURL(@blob)
        else
            @data = null
            @mdatadata?.content_type = undefined


    save: ->
        return unless @blob

        form = new FormData()
        form.append("id", @id) if @id
        form.append("file", @blob, @blob.name)
        form.append("owner_type", _.classify(@parent.modelTypeIdentifier()))
        form.append("owner_id", @parent.getId())
        form.append("owner_association", @parent_association)
        form.append("extension_id", @parent.extensionIdentifier())

        url = Lanes.config.api_path + '/asset'

        Lanes.Vendor.xhr.post(url, {body: form}, (err, resp, body) =>
            if err
                @errors = { http: err.message }
            else
                reply = JSON.parse(body)
                if reply.errors or reply.success is false
                    @errors = reply.errors
                else
                    @metadata = @blob = @data = null
                    @set( reply.data )
        )
