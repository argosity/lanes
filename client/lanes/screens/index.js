import { observable, autorun } from 'mobx';
import { map, uniq } from 'lodash';
import Group from './group';
import Config from '../config';
import Sync from '../models/sync';

import user from '../user';

const Screens = observable({

    all: observable.map(),

    active: observable.array(),

    groups: observable.map(),

    get activeGroups() {
        return uniq(map(this.active, s => Group.forId(s.group_id)));
    },

    refresh() {
        Sync.perform(`${Config.api_path}/lanes/screens`).then(({ data: { screens: screenIds } }) => {
            this.active.replace(map(screenIds, id => this.all.get(id)));
        });
    },

});
export default Screens;

let previousLoggedIn = null;
autorun(() => {
    if (user.isLoggedIn !== previousLoggedIn){
        previousLoggedIn = user.isLoggedIn;
        Screens.refresh();
    }
});
