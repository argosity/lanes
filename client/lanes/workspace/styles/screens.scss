.screen {
    padding: 5px;
    //  overflow: hidden;
    width: 100%;
    .toolbar {
        margin-top: 10px;
    }

    .toolbar{
        @extend .navbar;
        @extend .navbar-default;

        .btn {
            @extend .btn-default;
            @extend .navbar-btn;
        }

        .arrow {
            border-style: solid;

            border-left-color: transparent;
            border-right-color: transparent;

            border-bottom-color: $popover-title-bg;
            border-top-color: $popover-title-bg !important; 

            border-width: 0 9px 9px;
            height: 0;
            left: 7px;
            margin-top: 10px;
            opacity: 1;
            position: absolute;
            top: -18px;
            transition: all 0.25s ease 0s;
            width: 0;
            z-index: 10;
        }
        .navbar-left {
            float: left;
        }
        .navbar-right{
            float: right;
            .arrow {
                left: unset;
                right: 10px;
            }
        }

        .dropdown-toggle {
            .badge{
                padding: 2px 2px 0;
                position: relative;
                font-size: 70%;
                top: -6px;
                right: 5px;
            }
            position: relative;
            padding-right: 2px;
            i {
                /* position: absolute; */
                /* left: 0px; */
            }
        }

    }

}
