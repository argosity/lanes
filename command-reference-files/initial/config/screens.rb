# An example of how to define a screen group
# Hippo::Screen.define_group 'appy-app-screens' do | group |
#     group.title       = "AppyApp Screens"
#     group.description = "Screens relating to AppyApp"
#     group.icon        = "heart"
# end

Hippo::Screen.for_extension 'appy-app' do | screens |

end
