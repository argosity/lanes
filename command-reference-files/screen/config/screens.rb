# An example of how to define a screen group
# Hippo::Screen.define_group 'appy-app-screens' do | group |
#     group.title       = "AppyApp Screens"
#     group.description = "Screens relating to AppyApp"
#     group.icon        = "heart"
# end

Hippo::Screen.for_extension 'appy-app' do | screens |
    screens.define "ready-set-go" do | screen |
        screen.title       = "Ready Set Go"
        screen.description = ""
        screen.icon        = ""
        screen.group_id    = ""
        screen.model_class = ""
        screen.view_class  = "ReadySetGo"
    end

end
