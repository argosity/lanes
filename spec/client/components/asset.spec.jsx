import React from 'react';

import { Col } from 'react-flexbox-grid';
import Asset from 'lanes/components/asset';

import { TestImage } from '../test-models';
import { Snapshot } from 'lanes/testing/screens';

describe("Asset Component", () => {
    it('renders and matches snapshot', () => {
        const image = new TestImage();
        const asset = mount(<Asset model={image} name="asset" />);
        expect(asset).toHaveRendered('Dropzone');
        expect(Snapshot(<Asset model={image} name="asset" />)).toMatchSnapshot();
    });
});
