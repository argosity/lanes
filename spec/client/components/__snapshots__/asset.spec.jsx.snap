// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Asset Component renders and matches snapshot 1`] = `
<div
  className="asset form-field"
>
  <div
    className="grommetux-form-field grommetux-form-field--size-medium"
    onClick={[Function]}
  >
    <label
      className="grommetux-form-field__label"
      htmlFor={undefined}
    >
      Asset
    </label>
    <span
      className="grommetux-form-field__contents"
    >
      <div
        className="drop-zone"
        onClick={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        style={Object {}}
      >
        <div>
          Drop a file here, or click to select one to upload.
        </div>
        <input
          accept={undefined}
          multiple={false}
          onChange={[Function]}
          style={
            Object {
              "display": "none",
            }
          }
          type="file"
        />
      </div>
    </span>
  </div>
</div>
`;
